I18n - compile time inlining

petebacondarwin opened the pull request.

AndrewKushnir commented:
@petebacondarwin FYI I'm in the process of reviewing this PR, will let you know once it's completed. Thank you.

petebacondarwin commented:
Thanks Andrew. I am going to tweak the API a bit, today, after discussions with @alan-agius4. But I will add the changes as fixup commits so that it doesn't affect your review if you are going commit by commit.

AndrewKushnir commented on packages/localize/src/tools/src/translate/source_files/source_file_utils.ts:
```diff
@@ -0,0 +1,188 @@
+* Is the given `identifier` declared globally.
+* @param identifier The identifier to check.
+*/
+export function isGlobal(identifier: NodePath<t.Identifier>) {
```
nit: may be extend function name to include `Identifier` as well? Like `isGlobalIdentifier`?

AndrewKushnir commented on packages/localize/src/tools/src/translate/source_files/source_file_utils.ts:
```diff
@@ -0,0 +1,188 @@
+export function unwrapSubstitutionsFromLocalizeCall(call: t.CallExpression): t.Expression[] {
+  const expressions = call.arguments.splice(1);
+  if (!isArrayOfExpressions(expressions)) {
+    throw new Error('Invalid substitutions for `$localize` (expected an array of expressions).');
```
Can we add extra debug info here (based on info available in `call`) so that it's easier to find the necessary part of the code based on the error message?

petebacondarwin replied:
I have added better debug information - see these new tests:

https://github.com/angular/angular/pull/32881/files#diff-277f8bb6b90b1170f54e2cd51fe996baR90-R206

AndrewKushnir commented on packages/localize/src/tools/src/translate/source_files/source_file_utils.ts:
```diff
@@ -0,0 +1,188 @@
+*/
+export function unwrapStringLiteralArray(array: t.Expression): string[] {
+  if (!isStringLiteralArray(array)) {
+    throw new Error('Unexpected messageParts for `$localize` (expected an array of strings).');
```
Can we add extra debug info here as well?

AndrewKushnir commented on packages/localize/src/tools/src/translate/translation_files/translation_parsers/translation_parse_error.ts:
```diff
@@ -0,0 +1,30 @@
+
+  contextualMessage(): string {
+    const ctx = this.span.start.getContext(100, 3);
+    return ctx ? `${this.msg} ("${ctx.before}[${ParseErrorLevel[this.level]} ->]${ctx.after}")` :
```
👍

AndrewKushnir commented on packages/localize/src/tools/src/translate/translation_files/translation_parsers/xliff1/xliff1_message_serializer.ts:
```diff
@@ -0,0 +1,58 @@
+    this.renderer.closeContainer();
+    this.renderer.text(`}`);
+  }
+
```
Extra whitespace below.

AndrewKushnir commented on packages/localize/src/tools/test/translate/integration/BUILD.bazel:
```diff
@@ -0,0 +1,32 @@
+load("//tools:defaults.bzl", "jasmine_node_test", "ts_library")
```
I'm wondering if it makes sense to add e2e tests that would trigger all aspects of translation pipeline, for example:

- call `ng xi18n` to extract translations from a demo app (we can invoke it with different `--i18n-format` flags to generate all supported formats)
- update content of translation files (replace pre-defined string like `Hello` -> `Bonjour`, etc)
- run Ivy compile-time inlining
- run an app and verify that it contains translated string

The idea is to make sure VE extraction logic is always up-to-date with the way Ivy compile-time inlining works (i.e. we can parse files generated by `ng xi18n`). I noticed that we have tests for different formats (which are very useful!), but I'm wondering if we can have similar tests with "dynamic" input, i.e. content generated via `ng xi18n` (or via API functions).

Note: this is *not* a blocked, we can do that later if it makes sense.

petebacondarwin replied:
I will add this now.

petebacondarwin replied:
Done

AndrewKushnir reviewed (approved):
Very impressive work @petebacondarwin! 👍 
I left a few minor comments, none of them are blockers.
Thank you.

AndrewKushnir commented:
- [VE Presubmit](http://test/OCL:273417888:BASE:273417918:1570495783541:168edb0c)
- [Ivy Presubmit](http://test/OCL:273417888:BASE:273418066:1570495857899:83ccbbfd)

filipesilva commented on integration/cli-hello-world-ivy-i18n/angular.json:
```diff
@@ -40,7 +40,7 @@
               "optimization": true,
               "outputHashing": "all",
-              "sourceMap": false,
+              "sourceMap": true,
```
Why did sourcemaps need to be added? It doesn't look like they are tested.

petebacondarwin replied:
That is an artefact of trying to get source map merging working. I will revert this for now.

filipesilva reviewed (approved):
LGTM from dev-infra

petebacondarwin committed:
feat(ivy): i18n - implement compile-time inlining

This commit implements a tool that will inline translations and generate
a translated copy of a set of application files from a set of translation
files.

petebacondarwin committed:
test(ivy): i18n - add compile time translation to integration test

petebacondarwin committed:
refactor(ivy): i18n - rename `I18nError` to `TranslationParserError`

This closer reflects what caused the error.

AndrewKushnir commented:
FYI, VE and Ivy presubmits are successful. Thank you.

IgorMinar reviewed (approved):
I did a high level review of the PR (general architecture, public api, docs, tests) but didn't check all the details - I'm relying on other reviewers for that.

everything looks good, except for the missing license for babel__core types. Can you please fix that - you can create a new license.md file if one doesn't exist in the original repo, as long as you populate it with the correct licensing info. thanks!

IgorMinar commented:
hmm.. and ci seems to be failing due to duplicate dependencies? @petebacondarwin can you please take a look?

petebacondarwin committed:
fixup! feat(ivy): i18n - implement compile-time inlining

petebacondarwin commented on packages/localize/src/tools/types/babel/template.d.ts:
```diff
@@ -0,0 +1,78 @@
+  const templateBuilder: DefaultTemplateBuilder;
+
+  export default templateBuilder;
+}
```
I'm not sure how to configure the Bazel rules to do that...

petebacondarwin replied:
I'm going to ask @alxhub

alxhub replied:
I'm not either, unfortunately. That would be a good @alexeagle question. I don't know how `npm_package` determines the set of files to include.

Do references to these types end up in the `.d.ts` files for the package? If so, it might be _very_ difficult to get the package build to exclude them.

petebacondarwin replied:
Instead I suggest that we just try to get https://github.com/bazelbuild/rules_nodejs/issues/1033 fixed, which will allow us to remove these typings files altogether.

How about we land this PR now and then work to fix that issue and remove the typings before RC?

alxhub replied:
Per Igor, this is the approach we'll take. I will merge this PR soon.

kara replied:
@petebacondarwin Is there a Jira for this? Who is working on the Bazel fixes?

petebacondarwin replied:
@alexeagle is working on the fix as part of the Bazel team. I don't know if there is a JIRA for it. There is not one in the FW JIRA.

petebacondarwin committed:
fixup! feat(ivy): i18n - implement compile-time inlining

petebacondarwin committed:
fixup! feat(ivy): i18n - implement compile-time inlining

petebacondarwin commented:
@IgorMinar  - @babel/core typings LICENSE file added
